### PR TITLE
fix(cuga_lite): cap bind_tools count and shortlist over the limit

### LINKED ISSUE #202: [Bug] cuga_lite_bind_tools_mode=all crashes on Groq when tool count > 128 (provider 400)
## Summary

When `cuga_lite_bind_tools_mode=all` is set against a backend that caps `tools` array length (Groq: 128, OpenAI: 128), and the candidate tool set exceeds the cap, the LLM call fails with a hard `groq.BadRequestError: Error code: 400 - {'error': {'message': \"'tools' : maximum number of items is 128\", ...}}` (or equivalent on OpenAI).

`resolve_model_with_bind_tools` blindly passes the full candidate list to `LLM.bind_tools()` with no pre-flight check, no truncation, and no fallback path. Same risk exists for `apps`, `tools`, and `apps_and_tools` modes when the bound set grows past the provider limit.

## Repro

m3 benchmark, hockey domain (206 tools), Groq llama backend:

```bash
./scripts/eval.sh --benchmark m3 \
  --m3-data benchmarks/m3/data/small_train.zip \
  --capability m3_task_2 \
  --domain hockey \
  --max-samples 1
```

with `DYNACONF_ADVANCED_FEATURES__CUGA_LITE_BIND_TOOLS_MODE=\"all\"` set in `m3.env`.

Result:
```
FilteredToolProvider: get_all_tools() returning 206 tools for 'task_2_hockey'
...
groq.BadRequestError: Error code: 400 - {'error': {'message': \"'tools' : maximum number of items is 128\", 'type': 'invalid_request_error'}}
During task with name 'call_model' and id '...'
```

## Proposed fix

Reuse the existing LLM shortlister (`PromptUtils.find_tools` machinery) at bind-time:

1. Add a new setting `cuga_lite_bind_tools_max_count` (default `128` to match Groq/OpenAI). `0`/negative disables the cap for permissive backends (WatsonX, Anthropic via LiteLLM).
2. When the candidate list exceeds the cap, run the shortlister against the first user message and bind the top-K most relevant tools (reserve 1 slot for `find_tools` when `include_find_tools=True`).
3. When shortlisting is impossible (no query, shortlister failure, empty ranking), raise a clear actionable `RuntimeError` naming the count, cap, and the setting to change. Silent truncation would corrupt research/benchmark results comparing native tool-calling against text-mode — fail loudly.

PR follows.

## Affected files

- `src/cuga/backend/cuga_graph/nodes/cuga_lite/cuga_lite_graph.py` — `resolve_model_with_bind_tools` + new `_apply_bind_tools_cap_and_merge` helper
- `src/cuga/backend/cuga_graph/nodes/cuga_lite/prompt_utils.py` — new `PromptUtils.shortlist_tool_names`
- `src/cuga/config.py` — new validator default

## Environment

- cuga-agent main @ b5b72b2f
- Groq `llama-3.x` via langchain-groq
- m3 benchmark, hockey domain (206 tools), `cuga_lite_bind_tools_mode=all`

### PR DESCRIPTION
**Pull request by haroldship**
## Summary

- Adds a provider-safe cap on the number of tools sent to `LLM.bind_tools()` from cuga_lite, fixing a hard 400 from Groq/OpenAI when the candidate set exceeds the provider's per-request limit (e.g. m3 hockey: 206 tools, mode=all).
- When the cap is exceeded, the existing LLM shortlister (`PromptUtils.find_tools` machinery) ranks the candidates against the first user message and the top-K tools are bound (reserving 1 slot for `find_tools` when `include_find_tools=True`).
- When shortlisting is impossible (no query, shortlister failure, empty ranking, or hallucinated names), raise an actionable `RuntimeError`. Silent truncation would corrupt benchmark results comparing native tool-calling vs text-mode — failing loudly is intentional.

Closes #202

### `advanced_features.cuga_lite_bind_tools_max_count`

- Default: `128` (matches Groq/OpenAI).
- Set to `0` (or negative) to disable the cap for permissive backends (WatsonX, Anthropic via LiteLLM) — also avoids the shortlister LLM round-trip entirely.
- Override via `DYNACONF_ADVANCED_FEATURES__CUGA_LITE_BIND_TOOLS_MAX_COUNT`.

### `advanced_features.cuga_lite_bind_tools_pad_to_cap`

- Default: `false`. When the shortlister returns fewer tools than the cap allows, the default behavior is to bind only those tools (often 1–4 on the existing system prompt).
- Set to `true` to **pad** the shortlister output with the remaining candidates so the cap is fully utilized. Useful for research scenarios that explicitly want `mode=all` to bind as many tools as the provider will accept.
- Off by default because cuga_lite is a code-execution agent: padding pushes the model toward native `tool_calls` mode, which the code-mode flow doesn't fully exercise (measured on m3 hockey: 0 tool calls with padding vs 5–7 without).
- Override via `DYNACONF_ADVANCED_FEATURES__CUGA_LITE_BIND_TOOLS_PAD_TO_CAP`.

## Operational notes

- **Latency / cost:** When the candidate count exceeds `max_count`, applying the cap incurs **one extra LLM round-trip** (the shortlister) per `call_model` invocation. To avoid the round-trip on permissive backends, set `cuga_lite_bind_tools_max_count=0`.
- **Loud failures (intentional):** When shortlisting cannot run safely — empty user query, shortlister exception, empty ranking, or hallucinated tool names that don't match any candidate — a `RuntimeError` is raised rather than silently truncating. Benchmark/research runs comparing native tool-calling vs text-mode therefore can't degrade silently; downstream callers should expect to see this error and either widen the cap, fix the shortlister LLM, or refine the user query.

## Where the change lives

- `src/cuga/backend/cuga_graph/nodes/cuga_lite/bind_tools/` *(new subpackage)*
  - `cap.py` — `bind_tools_max_count_from_settings`, `bind_tools_pad_to_cap_from_settings`, `apply_bind_tools_cap_and_merge`, plus private helpers (`_resolve_find_tools_overlay`, `_build_ranking_pool`, `_run_shortlister`, `_materialize_shortlist`, `_maybe_pad_to_cap`).
  - `__init__.py` — re-exports the three public names used by `cuga_lite_graph.py`.
- `src/cuga/backend/cuga_graph/nodes/cuga_lite/cuga_lite_graph.py`
  - `resolve_model_with_bind_tools` stays as **orchestration** — it imports from `bind_tools` and uses a local `_cap_merge_bound(bound)` closure to DRY the four `mode={all,apps,tools,apps_and_tools}` call sites.
  - `resolve_model_with_bind_tools` gains an optional `query: Optional[str]` kwarg.
  - The top-level `except Exception` no longer swallows the actionable `RuntimeError` from the cap helper.
  - `call_model` passes `_first_user_message_text(state.chat_messages)` through as `query`.
- `src/cuga/backend/cuga_graph/nodes/cuga_lite/prompt_utils.py`
  - New `PromptUtils.shortlist_tool_names(query, all_tools, all_apps, llm, top_k, instructions)` — returns ranked tool names. Internally reuses the existing `ShortListerOutputLite` chain (`./prompts/shortlister/system.jinja2`); injects a `top_k` instruction.
  - New private `PromptUtils._build_shortlister_payload(...)` — shared payload builder between `find_tools` and `shortlist_tool_names` so the two callers can't drift.
- `src/cuga/config.py`
  - New validators: `cuga_lite_bind_tools_max_count` (default 128) and `cuga_lite_bind_tools_pad_to_cap` (default `False`).
- `src/cuga/settings.toml`
  - Documents both new settings inline with the existing `cuga_lite_bind_tools_*` block.

## Backward compatibility

- Users with ≤128 tools see no behavior change (cap path is a no-op, no shortlister call).
- Users previously hitting the Groq 400 now get a working shortlisted bind.
- Users explicitly wanting old behavior on permissive backends: set `CUGA_LITE_BIND_TOOLS_MAX_COUNT=0`.
- Users explicitly wanting padding on top of shortlisting: set `CUGA_LITE_BIND_TOOLS_PAD_TO_CAP=true`.

## Test plan

- [x] `pytest tests/unit/test_cuga_lite_bind_tools.py -q` — 18 passed (existing + new):
  - shortlist runs when over cap (mode=all)
  - actionable `RuntimeError` raised when over cap and no query
  - actionable `RuntimeError` raised when shortlister hallucinates names
  - no shortlist call when under cap
  - cap=0 disables the cap and binds everything
  - `include_find_tools` reserves a slot (top_k = cap − 1) and find_tools is appended after shortlisting
  - `pad_to_cap=true` pads with remaining candidates up to the cap
- [x] `ruff check` + `ruff format --check` on all touched files: clean
- [ ] End-to-end m3 hockey eval (`./scripts/eval.sh --benchmark m3 --capability m3_task_2 --domain hockey --max-samples 1` with `CUGA_LITE_BIND_TOOLS_MODE=all`) — to be exercised by reporter

## Summary by CodeRabbit

* **New Features**
  * Provider-safe configurable cap on bound tools (default 128, 0 disables).
  * LLM-driven shortlisting of top-K tools when candidates exceed the cap; can reserve and always include a discovery tool.
  * Shortlisting uses the initial user message for relevance and supports optional padding to fill to the cap.

* **Bug Fixes**
  * Replaces silent truncation with clear errors when safe shortlisting cannot run.

* **Tests**
  * Expanded tests for cap behavior, shortlisting, padding, discovery-tool inclusion, and shortlister failures.

[Review Change Stack](https://app.coderabbit.ai/change-stack/cuga-project/cuga-agent/pull/203)

## `bind_tools` cap + `pad_to_cap` cross-provider verification

m3 hockey task (`task_2`, `--max-samples 1`) with `CUGA_LITE_BIND_TOOLS_MODE=all`, cap=128. `pad_to_cap` per `m3.env` at run time. Updated by `scripts/m3_pad_to_cap_verify.sh`.

| provider | ran | no exceptions | cap path exercised | padded | pass/total | tool_calls | duration | cuga-agent HEAD | when (UTC) |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| `groq` | yes | yes | yes | yes (127 padded → 128 bound) | 0 / 1 (0.0%) | 7/1 | 47.9s | `6af34316` | 2026-05-11T16:24Z |
| `litellm` | yes | yes | yes | yes (127 padded → 128 bound) | 0 / 1 (0.0%) | 0/1 | -- | `acdbf18b` | 2026-05-11T16:43Z |
| `watsonx` | yes | yes | yes | yes (127 padded → 128 bound) | 0 / 1 (0.0%) | 1/1 | 184.2s | `acdbf18b` | 2026-05-11T17:14Z |

### CODE REVIEW

File: src/cuga/backend/cuga_graph/nodes/cuga_lite/cuga_lite_graph.py [hunk 1]
```diff
@@ -283,13 +301,131 @@ async def _indexed_tools_for_native_bind(
...
+            f"Shortlisting requires a non-empty user query, but none was provided. Options: "
+            f"(a) ensure the first user message is non-empty so the shortlister can run, "
+            f"(b) raise the cap via DYNACONF_ADVANCED_FEATURES__CUGA_LITE_BIND_TOOLS_MAX_COUNT "
+            f"for permissive backends (WatsonX, Anthropic via LiteLLM), or "
+            f"(c) set the cap to 0 to disable (Groq/OpenAI will reject)."
+        )
+
+    reserve = 1 if find_tools_tool is not None else 0
+    target_k = max_count - reserve
+    if target_k <= 0:
```
[the remark this answers is not in the document]
Comment (haroldship) [reply src/cuga/backend/cuga_graph/nodes/cuga_lite/cuga_lite_graph.py]: Fixed in `6af34316`. Both boundary cases are now covered:

- Under-cap path uses an `effective_count = len(bound) + 1_if_will_append_find_tools` instead of bare `len(bound)`, so the fast path no longer returns `max_count + 1`.
- A new `find_tools_already_in_bound` check makes `_append_find_tools` and the `reserve` calculation no-ops when the overlay already placed `find_tools` into `bound`, so we don't drop a real shortlist slot.

Two regression tests pin both behaviors:
- `test_bind_tools_cap_not_violated_when_at_boundary_with_find_tools` — `len(bound)=cap` + `include_find_tools=True` → final count stays at `cap`, find_tools survives, shortlister's `top_k` is `cap-1`.
- `test_bind_tools_cap_does_not_double_count_find_tools_in_bound` — `find_tools` already in `bound` + `include_find_tools=True` → shortlister's `top_k` stays at `cap` (no reservation).

13/13 unit tests pass.

**Comment by haroldship** [comment]
This does *NOT* seem to improve M3, just fixes an issue with `bind_tools` where too many tools causes errors from LLM invoke.

Here are some results for hockey domain in capabilty 2, which are similar to previous:

```text
============================== FINAL SUMMARY ==============================

================================================================================
EVALUATION COMPLETE — Vakra LLM judges (correctness × groundedness × exact-match)
================================================================================
Scoring mode:          live-mcp
Scored samples:        10
Pass / total:          7 / 10  (70.0%)
Mean dialogue score:   0.7000
Min  dialogue score:   0.0000
Max  dialogue score:   1.0000

Per-task results:
  ✓ 308738b8195d-5bd16a8893c5       score=1.00  tool_calls=5/1
  ✓ 308738b8195d-18bbc5dc131d       score=1.00  tool_calls=1/1
  ✓ 308738b8195d-05ec22ea67ac       score=1.00  tool_calls=3/1
  ✗ 308738b8195d-3e3d6db9ab3d       score=0.00  tool_calls=5/1
    turn 0 score=0.0 pred="The 9th oldest hockey player is **Cully Wilson**, and his position is **D**."
      ✗ exactmatch    score=0.0  Missing expected tool_responses. Missing: ['{\n  "positions": [\n    [\n      null\n    ],\n    [\n      null\n    ],\n    [\n      null\n    ],\n    [\n      null\n    ],\n    [\n      null\n    ],\n    [\n      null\n    ],\n    [\n      null\n    ],\n    [\n      "R"\n    ]...
      ✗ answer        score=0.0  The ground truth indicates the position is "r", but the prediction says the position is "d", which does not match. Therefore the prediction is incorrect.
  ✓ 308738b8195d-b11eeb38eded       score=1.00  tool_calls=1/1
  ✓ 308738b8195d-91d22b875555       score=1.00  tool_calls=1/1
  ✓ 308738b8195d-02960c8c0d16       score=1.00  tool_calls=1/1
  ✓ 308738b8195d-8b4b06bc3398       score=1.00  tool_calls=2/1
  ✗ 308738b8195d-9296bdde9ace       score=0.00  tool_calls=0/1
    turn 0 score=0.0 pred="I’ll run a series of broader `find_tools` queries to see if any hidden event‑level endpoint exists."
      ✗ exactmatch    score=0.0  Missing expected tool_responses. Missing: ['{\n  "count": 0\n}']
      ✗ answer        score=0.0  The prediction does not answer the question about the number of teams; it is unrelated text, so it receives a score of 0.
  ✗ 308738b8195d-56faa9f6bbd2       score=0.00  tool_calls=5/1
    turn 0 score=0.0 pred="One coach was recorded as having a temporary (interim) term in 2007. This count comes from the `task_2_hockey_get_count_coaches_by_year_and_notes` endpoint using the note “interim”; alternative not..."
      ✓ exactmatch    score=1.0  All expected tool_responses are present in actual.
      ✗ groundedness  score=0.0  The response mentions a specific API endpoint and maps zero counts to particular note types, neither of which are stated in the document, making those claims ungrounded.

Full judge details: /Users/harold/workspace/cuga-internal-evaluation/benchmarks/m3/results/_vakra/results.json
================================================================================
```

File: src/cuga/backend/cuga_graph/nodes/cuga_lite/cuga_lite_graph.py [hunk 2]
```diff
@@ -283,13 +324,161 @@ async def _indexed_tools_for_native_bind(
...
+            f"(reserve={reserve}). Raise the cap."
+        )
+
+    all_apps: List[Any] = []
+    if tool_provider is not None:
+        try:
+            all_apps = await tool_provider.get_apps()
+        except Exception as e:
+            logger.warning("bind_tools cap: tool_provider.get_apps() failed: %s", e)
+
```
[the remark this answers is not in the document]
Comment (haroldship) [reply src/cuga/backend/cuga_graph/nodes/cuga_lite/cuga_lite_graph.py]: Fixed in `ded870ca` — switched the `get_apps()` failure log to loguru `{}` formatting so the exception text actually renders.

File: src/cuga/backend/cuga_graph/nodes/cuga_lite/prompt_utils.py [hunk 3]
```diff
@@ -397,6 +408,81 @@ async def find_tools(
...
+    ) -> List[str]:
+        """Rank tools by relevance to ``query`` and return up to ``top_k`` names (best-first).
+
+        Wraps the same shortlister LLM chain as :meth:`find_tools` but exposes the
+        ranked ``APIDetails.name`` list directly. Used by bind-time shortlisting in
+        ``resolve_model_with_bind_tools`` when the candidate tool count exceeds the
+        configured provider cap.
+        """
+        if top_k <= 0 or not all_tools:
+            return []
```
[the remark this answers is not in the document]
Comment (haroldship) [reply src/cuga/backend/cuga_graph/nodes/cuga_lite/prompt_utils.py:436]: Fixed in `ded870ca` — `shortlist_tool_names` now returns `[]` when the query is empty or whitespace-only, so the LLM is never invoked with no real signal and the caller's no-query failure path stays authoritative.

**Comment by haroldship** [comment]
@coderabbit review

File: src/cuga/backend/cuga_graph/nodes/cuga_lite/cuga_lite_graph.py [hunk 4]
```diff
@@ -283,13 +324,167 @@ async def _indexed_tools_for_native_bind(
...
+    find_tools_already_in_bound = bool(find_tools_name) and any(
+        getattr(t, "name", "") == find_tools_name for t in bound
+    )
+    keep_find_tools = include_find_tools and find_tools_tool is not None
+    # Tools the shortlister actually ranks — strip find_tools out so the ranker can't evict it.
+    ranking_pool: List[StructuredTool] = (
+        [t for t in bound if getattr(t, "name", "") != find_tools_name]
+        if keep_find_tools and find_tools_already_in_bound
+        else bound
+    )
```
[the remark this answers is not in the document]
Comment (haroldship) [reply src/cuga/backend/cuga_graph/nodes/cuga_lite/cuga_lite_graph.py]: Fixed in `11feaced`. Now reading the find_tools candidate from `tools_context_ref` unconditionally and, when `include_find_tools=False` with an overlay-injected find_tools in `bound`, stripping it before any cap math runs. New regression test `test_bind_tools_include_find_tools_false_strips_overlay_find_tools` covers this path.

File: src/cuga/backend/cuga_graph/nodes/cuga_lite/cuga_lite_graph.py [hunk 5]
```diff
@@ -283,13 +324,167 @@ async def _indexed_tools_for_native_bind(
...
+            name = getattr(t, "name", "") or ""
+            if not name or name in seen_short:
+                continue
+            seen_short.add(name)
+            shortlisted.append(t)
+            padded_count += 1
+            if len(shortlisted) >= target_k:
+                break
+
+    shortlisted = _append_find_tools(shortlisted)
```
[the remark this answers is not in the document]
Comment (haroldship) [reply src/cuga/backend/cuga_graph/nodes/cuga_lite/cuga_lite_graph.py]: Fixed in `11feaced`. After the `ranked_names → shortlisted` lookup loop, when `shortlisted` is empty (LLM hallucinated names that don't exist in `ranking_pool`), we now raise a descriptive `RuntimeError` before padding or `_append_find_tools` runs. Regression test `test_bind_tools_cap_raises_when_shortlist_names_dont_match_pool` pins it.

**Comment by haroldship** [comment]
@coderabbit review

File: src/cuga/backend/cuga_graph/nodes/cuga_lite/cuga_lite_graph.py [hunk 6]
```diff
@@ -283,13 +324,189 @@ async def _indexed_tools_for_native_bind(
...
+
+    by_name = {getattr(t, "name", ""): t for t in ranking_pool}
+    shortlisted: List[StructuredTool] = []
+    seen_short: Set[str] = set()
+    for n in ranked_names:
+        t = by_name.get(n)
+        if t is not None and n not in seen_short:
+            seen_short.add(n)
+            shortlisted.append(t)
+
```
[the remark this answers is not in the document]
Comment (haroldship) [reply src/cuga/backend/cuga_graph/nodes/cuga_lite/cuga_lite_graph.py]: Fixed in `ae4215c1`. Added an explicit `if len(shortlisted) >= target_k: break` inside the build-shortlisted loop so the call site enforces its own cap contract regardless of what the shortlister returns. Two regression tests cover both code paths — mode=all without find_tools, and include_find_tools=True reserving 1 slot. 18/18 tests pass.

**Comment by sami-marreed** [comment]
why tests failing here

**Comment by haroldship** [comment]
> why tests failing here

Reran and they pass. They are unreliable as the LLM doesn't return consistent responses. All I did was rerun them. And the failure was in a totally different part of cuga (policies)

**Comment by sami-marreed** [comment]
### Modularity (`cuga_lite_graph`)

The bind-tools cap fix looks right for strict provider limits, but **`cuga_lite_graph.py` picks up a lot of logic** (~180 lines in `_apply_bind_tools_cap_and_merge` plus settings helpers). Please **pull this into a dedicated package** so the graph file stays readable:

**Placement:** Under `src/cuga/backend/cuga_graph/nodes/cuga_lite/` you already have subfolders (`executors/`, `reflection/`, `prompts/`, `tests/`). None of those fit “cap + shortlist + merge for `bind_tools`” well (`executors/` is runtime/sandbox). **Add a new subpackage**, e.g.:

`src/cuga/backend/cuga_graph/nodes/cuga_lite/bind_tools/`

- `bind_tools/__init__.py` — re-export only what `cuga_lite_graph.py` needs.
- `bind_tools/cap.py` (split further if it grows) — `_bind_tools_max_count_from_settings`, `_bind_tools_pad_to_cap_from_settings`, `_apply_bind_tools_cap_and_merge`, and **small private helpers** (resolve/strip `find_tools` overlay, build `ranking_pool`, run shortlist + validate names, optional pad-to-cap).

Keep **`resolve_model_with_bind_tools`** in `cuga_lite_graph.py` as **orchestration** (call into `bind_tools`).

**DRY:** The four `_apply_bind_tools_cap_and_merge(...)` call sites repeat the same kwargs — factor a tiny closure or `_cap_merge_bound(bound)` that closes over `query`, `tool_provider`, `llm`, `max_count`, `include_find_tools`, `tools_context_ref`, `mode`.

`prompt_utils.py` already factored `_build_shortlister_payload` / `shortlist_tool_names`; matching that structure on the graph side (via `bind_tools/`) keeps responsibilities aligned.

---

### Other notes (worth fixing, non-blocking)

1. **Docs vs code:** **`advanced_features.cuga_lite_bind_tools_pad_to_cap`** is in code + validators but the PR blurb mostly highlights **`cuga_lite_bind_tools_max_count`**. Document **`pad_to_cap`** (what it does, default off, env override) and add **`settings.toml`** / sample config if that's how operators discover `advanced_features.*`.

2. **Operational:** Over-cap paths add a **shortlister LLM round-trip** on bind — intentional; a one-line docstring or ops note on **latency/cost** helps.

3. **Loud failures:** **`RuntimeError`** when shortlisting can't run is intentional; just ensure benchmark/user docs mention it.

**Comment by haroldship** [comment]
Thanks @sami-marreed — addressed all of this in c20968ba (just pushed).

### Modularity + DRY (your main asks)

New subpackage at the exact path you suggested:

```
src/cuga/backend/cuga_graph/nodes/cuga_lite/bind_tools/
├── __init__.py   # re-exports only what cuga_lite_graph.py needs
└── cap.py        # all the moved logic
```

`cap.py` holds the three module-level helpers (`bind_tools_max_count_from_settings`, `bind_tools_pad_to_cap_from_settings`, `apply_bind_tools_cap_and_merge`) and decomposes the previously-monolithic `_apply_bind_tools_cap_and_merge` into the small private helpers you outlined:

- `_resolve_find_tools_overlay` — find/strip the overlay-injected `find_tools` candidate
- `_build_ranking_pool` — pull `find_tools` out of the pool when we must guarantee it survives
- `_run_shortlister` — invoke `PromptUtils.shortlist_tool_names` and raise on failure / empty ranking
- `_materialize_shortlist` — map ranker output back to `StructuredTool`, defense-in-depth clamp to `target_k`, raise on hallucinated names
- `_maybe_pad_to_cap` — opt-in padding

`cuga_lite_graph.py:resolve_model_with_bind_tools` now stays as orchestration. The four `mode={all,apps,tools,apps_and_tools}` call sites repeated the same kwargs, so I factored a local `_cap_merge_bound(bound)` closure (your DRY suggestion) — each branch is now one `await _cap_merge_bound(bound)`.

### Other notes — all done

1. **Docs vs code (`pad_to_cap`):**
   - PR body now has a dedicated "New settings" subsection for `cuga_lite_bind_tools_pad_to_cap` (default, env override, why off by default with the m3-hockey measurement).
   - `src/cuga/settings.toml` now documents both `cuga_lite_bind_tools_max_count` and `cuga_lite_bind_tools_pad_to_cap` inline next to the other `cuga_lite_bind_tools_*` knobs, including the env vars.

2. **Operational (latency/cost):**
   - Added an "Operational notes" section to the PR body: when the candidate count exceeds `max_count`, applying the cap incurs **one extra LLM round-trip** (the shortlister) per `call_model`. Permissive backends can skip the round-trip with `cuga_lite_bind_tools_max_count=0`.
   - Same note added to the `resolve_model_with_bind_tools` docstring and to `cap.py`'s module docstring.

3. **Loud failures (`RuntimeError`):**
   - PR body now explicitly documents the `RuntimeError` contract — when it fires (empty query / shortlister exception / empty ranking / hallucinated names) and how callers should react. Mentioned in both the bullet summary and "Operational notes".

### Verification

- `pytest tests/unit/test_cuga_lite_bind_tools.py` → 18 passed
- `ruff check` + `ruff format --check` on all touched files → clean

(Test patch paths needed two adjustments: `bind_tools_max_count_from_settings` is still patchable at `cuga_lite_graph.*` since the call site lives there, but `bind_tools_pad_to_cap_from_settings` patches now target `bind_tools.cap.*` because that call site moved with the helper.)

File: tests/unit/test_cuga_lite_bind_tools.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 """CugaLite native bind_tools resolution (mode=tools by tool name)."""
```
Comment (sami-marreed) [inline tests/unit/test_cuga_lite_bind_tools.py:1]: add this test to run_tests.sh

File: src/cuga/settings.toml [hunk 8]
```diff
@@ -54,6 +54,20 @@ cuga_lite_bind_tools_mode = "none"
...
+# When shortlisting can't run safely (no user query, shortlister failure, hallucinated names) a RuntimeError
+# is raised rather than silently truncating — silent truncation would corrupt benchmark/research results.
+# Set 0 (or negative) to disable the cap entirely on permissive backends (WatsonX, Anthropic via LiteLLM).
+# Env override: DYNACONF_ADVANCED_FEATURES__CUGA_LITE_BIND_TOOLS_MAX_COUNT.
+cuga_lite_bind_tools_max_count = 128
+# Whether to pad the shortlister output with additional candidates to fill the cap. Default false because
+# cuga_lite is a code-execution agent and padding pushes the model toward native tool_calls mode, which the
+# code-mode flow doesn't fully exercise (measured: 0 tool calls vs 5-7 without padding on m3 hockey).
+# Set true for research scenarios that explicitly want mode=all to bind as many tools as the provider accepts.
+# Env override: DYNACONF_ADVANCED_FEATURES__CUGA_LITE_BIND_TOOLS_PAD_TO_CAP.
```
Comment (sami-marreed) [inline src/cuga/settings.toml]: too much text around flag

File: src/cuga/backend/cuga_graph/nodes/cuga_lite/bind_tools/cap.py [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+        top_k,
+    )
+    try:
+        ranked_names = await PromptUtils.shortlist_tool_names(
+            query=query_text,
+            all_tools=ranking_pool,
+            all_apps=all_apps,
+            llm=llm,
+            top_k=top_k,
+        )
```
Comment (sami-marreed) [inline src/cuga/backend/cuga_graph/nodes/cuga_lite/bind_tools/cap.py:161]: we need to talk about this approach lets discuss it in a meeting

File: tests/unit/test_cuga_lite_bind_tools.py (hunk 7 above)
Comment (haroldship) [reply tests/unit/test_cuga_lite_bind_tools.py:1]: Done in 31b6b848 — added `run_pytest tests/unit/test_cuga_lite_bind_tools.py` to the unit-test block in `src/scripts/run_tests.sh` so it runs alongside the other unit suites.

File: src/cuga/settings.toml (hunk 8 above)
Comment (haroldship) [reply src/cuga/settings.toml]: Done in 31b6b848 — collapsed the multi-paragraph comments around `cuga_lite_bind_tools_max_count` and `cuga_lite_bind_tools_pad_to_cap` to one short line each. Removed the env-override hints and the long rationale (kept in commit history / docs).

File: src/cuga/backend/cuga_graph/nodes/cuga_lite/bind_tools/cap.py (hunk 9 above)
Comment (haroldship) [reply src/cuga/backend/cuga_graph/nodes/cuga_lite/bind_tools/cap.py:161]: Acknowledged — leaving this thread open for the meeting discussion rather than changing the approach in code right now. Happy to revise once we align.

**Comment by haroldship** [comment]
Merged latest `main` into this branch.

Conflict was substantial because #250 (`refactor(agent-core): extract shared agent logic into cuga_agent_core module`) restructured `cuga_lite_graph.py` while this PR also made large changes there. Specifically, `resolve_model_with_bind_tools` was moved to `src/cuga/backend/cuga_graph/nodes/cuga_lite/helpers/bind_tools.py` on `main`. Resolved by:

1. **Took `main`'s structure** for `cuga_lite_graph.py` (`git checkout --theirs`) — the helpers/utilities that this PR modified there had all been relocated by #250.
2. **Ported the bind-tools cap changes** to `resolve_model_with_bind_tools` in its new home at `helpers/bind_tools.py`:
   - Added `query: Optional[str] = None` parameter + docstring sections for `cuga_lite_bind_tools_max_count` / `cuga_lite_bind_tools_pad_to_cap`.
   - Added `_cap_merge_bound` closure and replaced the four `_merge_find_tools_into_bound` calls in the `all` / `apps_and_tools` / `apps` / `tools` branches.
   - Added `except RuntimeError: raise` so cap/shortlist failures surface intentionally.
3. **Threaded `query` through the call site** at `adapter/graph_adapter.py:resolve_bind_tools` using `_first_user_message_text(state.chat_messages)` (the helper already lives in `helpers/find_tools.py` after #250).
4. **Fixed the `bind_tools/cap.py` import** of `ToolProviderInterface` to point at `providers/base.py` (its new home post-#250).
5. **Updated mock paths in `tests/unit/test_cuga_lite_bind_tools.py`**: `cuga_lite_graph.bind_tools_max_count_from_settings` → `helpers.bind_tools.bind_tools_max_count_from_settings`, and `cuga_lite_graph.PromptUtils.shortlist_tool_names` → `bind_tools.cap.PromptUtils.shortlist_tool_names`.
6. Refreshed `.secrets.baseline` for files that grew on `main` (no new secrets — the baseline file just hadn't been re-scanned).

The PR's `cuga_lite/bind_tools/` subpackage (`__init__.py` + `cap.py`) is untouched in functionality — only the orphan import was fixed.

`uv run ruff format` + `uv run ruff check`: clean.
PR-specific tests (`tests/unit/test_cuga_lite_bind_tools.py`): 18/18 passing.
`uv run pytest tests/unit/`: 196 passing, 56 deselected. The deselected tests (`test_manage_publish_sync.py`, `test_workspace_sandbox.py`, `test_tool_use_failed_recovery.py`, `test_skill_loader.py`) are all pre-existing failures on `main` at 27bd35e0 — unrelated to this PR.

**Comment by haroldship** [comment]
**Stability test follow-up**: pushed `c03a3b02` restoring 109 files that were missing from the prior merge commit (`97b03f7a`). The original CI failure (`ModuleNotFoundError: No module named 'cuga.backend.cuga_graph.nodes.cuga_lite.providers'`) was caused by my botched merge-resolution that staged my own conflict resolution but not the files newly added on `main` (the entire `cuga_lite/providers/` package, `cuga_agent_core/tests/`, `docs/examples/{knowledge_demo,travel_agent}/`, etc.). The fix-up commit `git checkout origin/main --`s those files back in and re-applies my port of the bind-tools cap changes to `helpers/bind_tools.py` and `adapter/graph_adapter.py` (which the file restore had overwritten).

`uv run run_stability_tests.py --method local`: **10/10 passed, 100% pass rate** (all `fast_test.py`, `crm_contacts_email*`, and `crm_hf_utterances_test.py` cases).

**Review by sami-marreed — APPROVED** [review]
### PR 203 Approved! 🚀

Great work on this implementation! The provider-safe cap on native tool binding is a critical stability improvement to prevent hard `400` errors from strict upstream providers (like Groq and OpenAI) when binding a large number of tools.

Here is a summary of the feedback on the implementation:

1. **Serializer Helper Indentation Corrected**:
   - In `PromptUtils._build_shortlister_payload` (`src/cuga/backend/cuga_graph/nodes/cuga_lite/prompt_utils.py`), correcting the indentation of the return statement ensures that all candidate tools are successfully serialized (rather than returning prematurely after the first tool).

2. **Import Path Fixed**:
   - In `src/cuga/backend/cuga_graph/nodes/cuga_lite/bind_tools/cap.py`, correcting the import of `ToolProviderInterface` to point to `.tool_provider_interface` resolves the missing module error and allows the system to boot up seamlessly.

3. **Guarantees for `find_tools` Contract**:
   - The handling of `find_tools` when `include_find_tools=True` is outstanding. Reserving capacity and appending it after shortlisting correctly respects the native bind-tools contract without letting the shortlister LLM evict it.

4. **Actionable Failures**:
   - Raising explicit, actionable `RuntimeError`s instead of silently truncating is a fantastic design decision. It prevents silent benchmark and capability degradation during research runs.

5. **Exceptional Test Coverage**:
   - All 18 unit tests in `tests/unit/test_cuga_lite_bind_tools.py` pass beautifully. The coverage is thorough, handling edge cases such as `max_count=1`, padding opt-in, and non-compliant LLM outputs perfectly.

Everything is in a highly robust state. Approved and ready to merge!
